File: listings/pocketledger/PocketLedger/Gate/GateService.swift
import Foundation

/// B 面打开方式：internal=内开（WKWebView）/ external=外开（系统浏览器）。
/// 仅 mode=B 时有意义，服务端只在判 B 时下发；缺省/非法一律内开。
enum GateOpenMode {
    case internalWebView
    case externalBrowser

    /// 只认 "external"，其余（含缺省 nil）一律内开。
    static func from(_ raw: String?) -> GateOpenMode {
        return raw == "external" ? .externalBrowser : .internalWebView
    }
}

/// 网关判定结果。默认即 A 面——任何不确定都落到这个安全默认值。
enum GateDecision {
    case aSide
    case bSide(url: URL, openMode: GateOpenMode)
}

/// 调用服务端 AB 面网关。判定全在服务端（按请求真实 IP 查国家 + 时区/IP 规则）；
/// 客户端只上报 platform/bundleId/timezone，从不上报 IP，也从不内置 B 面地址。
///
/// 安全不变量：**任何异常（超时、网络错误、非 200、解析失败）一律 A 面**。
enum GateService {
    /// 逐个尝试候选基址，任一成功即返回；全部失败 → A 面。
    static func evaluate(timezone: String) async -> GateDecision {
        let payload: [String: String] = [
            "platform": GateConfig.platform,
            "bundleId": GateConfig.bundleId,
            "timezone": timezone
        ]
        guard let body = try? JSONSerialization.data(withJSONObject: payload) else {
            return .aSide
        }

        for base in GateConfig.apiBases {
            if let decision = await post(base: base, body: body) {
                return decision
            }
        }
        return .aSide
    }

    /// 向单个基址发一次判定请求。返回 nil = 该基址连不上（外层继续试下一个）；
    /// 返回 .aSide = 拿到了明确的「非 B」响应（或任何应判 A 的情形）。
    private static func post(base: String, body: Data) async -> GateDecision? {
        guard let url = URL(string: base + GateConfig.gatePath) else { return nil }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = body
        req.timeoutInterval = GateConfig.requestTimeout

        do {
            let (data, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse else { return .aSide }
            guard http.statusCode == 200 else { return .aSide }
            return parse(data)
        } catch {
            // 连接类错误：让外层尝试下一个候选基址。其余无法区分时也回退 A 面。
            let code = (error as? URLError)?.code
            if code == .cannotConnectToHost || code == .cannotFindHost
                || code == .notConnectedToInternet || code == .timedOut {
                return nil
            }
            return .aSide
        }
    }

    /// 解析响应；形状不符一律 A 面。
    private static func parse(_ data: Data) -> GateDecision {
        guard
            let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            (obj["mode"] as? String) == "B",
            let urlString = obj["url"] as? String,
            let url = usableURL(urlString)
        else {
            return .aSide
        }
        return .bSide(url: url, openMode: GateOpenMode.from(obj["openMode"] as? String))
    }

    /// B 面 url 是否可用：能解析、是 http/https、且有主机名。
    ///
    /// 只判 `URL(string:)` 非 nil 是不够的 —— `URL(string: "somestring")` 会成功，
    /// 得到一个只有 path、没有 scheme 与 host 的相对 URL，WKWebView 加载它只会白屏。
    /// 白屏比回退 A 面糟得多（用户看到的是个坏掉的 App），所以这里前置挡掉：
    /// 挡不住的畸形值宁可判 A。与 colorstack / calcpad / tilefit 的 Dart 侧同口径。
    private static func usableURL(_ raw: String) -> URL? {
        guard !raw.isEmpty, let url = URL(string: raw) else { return nil }
        guard let scheme = url.scheme?.lowercased(), scheme == "http" || scheme == "https" else {
            return nil
        }
        guard let host = url.host, !host.isEmpty else { return nil }
        return url
    }
}

File: listings/reptimer/lib/gate/gate_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'gate_config.dart';

/// B 面 url 是否可用：能解析、且是 http/https、且有主机名。
///
/// 对齐 decktallypro `GateService.parse` 里的 `URL(string:)` 校验 —— 那边解析失败即判 A。
/// Flutter 的 `Uri.tryParse` 比 iOS 宽松（'not a url' 也能解析成一个只有 path 的 Uri），
/// 故这里额外要求 scheme 与 host，免得把一个 WebView 加载不了的地址当成有效 B 面。
bool _isUsableUrl(String raw) {
  if (raw.isEmpty) return false;
  final uri = Uri.tryParse(raw);
  if (uri == null) return false;
  return (uri.scheme == 'http' || uri.scheme == 'https') && uri.host.isNotEmpty;
}

/// 网关判定结果。默认构造即「A 面」——任何不确定都应落到这个安全默认值上。
class GateResult {
  const GateResult.aSide() : mode = 'A', url = null, openMode = 'internal';
  const GateResult.bSide(this.url, {this.openMode = 'internal'}) : mode = 'B';

  final String mode;
  final String? url;

  /// B 面打开方式：'internal'=内开（原生 WebView）/ 'external'=外开（外部浏览器）。
  /// 仅 mode=B 时有意义，服务端只在判 B 时下发；缺省/非法一律 internal（默认内开）。
  final String openMode;

  bool get isBSide => mode == 'B' && url != null && _isUsableUrl(url!);

  /// 是否外开（外部浏览器）。仅在 isBSide 时才应参考此值。
  bool get isExternal => openMode == 'external';
}

/// 调用服务端 AB 面网关。判定全在服务端（按请求真实 IP 查国家 + 时区/IP 规则）；
/// 客户端只上报 platform/bundleId/timezone，从不上报 IP，也从不内置 B 面地址。
///
/// 安全不变量：**任何异常（超时、网络错误、非 200、解析失败）一律返回 A 面**。
/// 与后端 fail-closed 一致——判错成 A 只是少放一个用户，判错成 B 可能放进审核员。
class GateService {
  const GateService();

  Future<GateResult> evaluate({required String timezone}) async {
    final body = jsonEncode(<String, String>{
      'platform': GateConfig.platform,
      'bundleId': GateConfig.bundleId,
      'timezone': timezone,
    });

    // 依次尝试各候选基址，任一成功即返回；全部失败落到 A 面。
    for (final base in GateConfig.apiBases) {
      final result = await _post(base, body);
      if (result != null) return result;
    }
    return const GateResult.aSide();
  }

  /// 向单个基址发一次判定请求。返回 null 表示这个基址不可用（外层继续试下一个）。
  Future<GateResult?> _post(String base, String body) async {
    final uri = Uri.tryParse('$base${GateConfig.gatePath}');
    if (uri == null) return null;

    final client = HttpClient()..connectionTimeout = GateConfig.requestTimeout;
    try {
      final req = await client.postUrl(uri).timeout(GateConfig.requestTimeout);
      req.headers.contentType = ContentType.json;
      req.add(utf8.encode(body));

      final resp = await req.close().timeout(GateConfig.requestTimeout);
      if (resp.statusCode != HttpStatus.ok) {
        return const GateResult.aSide();
      }
      final text = await resp
          .transform(utf8.decoder)
          .join()
          .timeout(GateConfig.requestTimeout);
      return _parse(text);
    } on TimeoutException {
      return const GateResult.aSide();
    } on SocketException {
      // 连不上这个基址：返回 null 让外层尝试下一个候选。
      return null;
    } catch (_) {
      return const GateResult.aSide();
    } finally {
      client.close(force: true);
    }
  }

  /// 解析响应体。形状不符一律当 A 面。
  GateResult _parse(String text) {
    try {
      final data = jsonDecode(text);
      if (data is Map && data['mode'] == 'B') {
        final url = data['url'];
        // 非空之外还必须真的能解析成 http(s) 地址，对齐 decktallypro 的
        // `URL(string:)` 校验（解析不出就判 A）。colorstack 只校验非空，一个畸形
        // url 会让内开路径的 Uri.parse 抛 FormatException、B 面白屏；这里前置挡掉，
        // 挡不住的畸形值宁可判 A 也不带着崩。
        if (url is String && _isUsableUrl(url)) {
          // openMode 只认 'external'，其余（含缺省）一律 internal（默认内开）。
          final openMode = data['openMode'] == 'external'
              ? 'external'
              : 'internal';
          return GateResult.bSide(url, openMode: openMode);
        }
      }
    } catch (_) {
      // 落到 A 面。
    }
    return const GateResult.aSide();
  }
}
